odl/contrib/fom/supervised.py

```diff
@@ -473,20 +497,47 @@ def ssim(data, ground_truth,
         Input data to compare to the ground truth.
     ground_truth : `FnBaseVector`
         Reference to compare ``data`` to.
-    size : odd int
+    size : odd int, optional
         Size in elements per axis of the Gaussian window that is used
         for all smoothing operations.
-    sigma : positive float
+    sigma : positive float, optional
         Width of the Gaussian function used for smoothing.
-    K1, K2 : positive float
-        TODO
-    dynamic_range : nonnegative float
-        TODO
+    K1, K2 : positive float, optional
+        Small constants to stabilize the result. See [Wan+2004] for details.
+    dynamic_range : nonnegative float, optional
+        Difference between the maximum and minimum value that the pixels
+        can attain. Use 255 if pixel range is :math:`[0, 255]` and 1 if
+        it is :math:`[0, 1]`. Default: `None`, obtain maximum and minimum
+        from the ground truth.
+    normalized  : bool, optional
+        If ``True``, the output values are mapped to the interval
+        :math:`[0, 1]` (see `Notes` for details), otherwise return the
+        original SSIM.
+    force_lower_is_better : bool, optional
+        If ``True``, it is ensured that lower values correspond to better
+        matches by returning the negative of the SSIM, otherwise the (possibly
+        normalized) SSIM is returned. If both `normalized` and
+        `force_lower_is_better` are ``True``, then the order is reversed before
+        mapping the outputs, so that the latter are still in the interval
+        :math:`[0, 1]`.
 
     Returns
     -------
     ssim : float
-        FOM value, where a higher value means a better match.
+        FOM value, where a higher value means a better match
+        if `force_lower_is_better` is ``False``.
+
+    Notes
+    -----
+    The unnormalized values are in the interval :math:`[-1, 1]`, where 1
+    corresponds to a perfect match. The normalized values are obtained by
```

Comment:
what does -1 correspond to?

Reply:
Does not seem to be so easy … first, the constants K1 and K2 (which prevent from dividing by zero) have to be set to zero, and even then the smallest possible value can only be achieved in special situations.

Reply:
Well in general there are basically two cases

either

* d(x, 0)

or 

* d(x, -x)

is minimal. In this case I guess that -1 corresponds to the second case?

Reply:
Checked this with Spyder. Some examples resulting from our implementation:

odl.contrib.fom.ssim([1], [-1])
Out[3]: 1.0

odl.contrib.fom.ssim([1], [1])
Out[4]: 1.0

odl.contrib.fom.ssim([1, -1], [-1, 1])
Out[5]: 0.57007198720979002

odl.contrib.fom.ssim([1, 0], [0, 1])
Out[6]: -0.080178871240947275

odl.contrib.fom.ssim([1, 0, 1], [0, 1, 0])
Out[7]: -0.16322592107522449

odl.contrib.fom.ssim([1, -1, 1], [-1, 1, -1])
Out[8]: 0.87354822079002448

odl.contrib.fom.ssim([1, -1, 1], [0, 0, 0])
Out[9]: 0.0

Reply:
I guess we'll leave this for now then since this seems complicated